sql: make all aggregate functions behave consistently.

knz opened the pull request:
All aggregate functions except for COUNT should return NULL if the source table has
zero rows, regardless of their argument. This means they cannot be
simplified to a constant Datum during normalization even if their argument
is constant.

Fixes #7039.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7043)

<!-- Reviewable:end -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> #### sql: make all aggregate functions behave consistently\.
> 
> All aggregate functions except for COUNT should return NULL if the source table has
> 
> zero rows, regardless of their argument. This means they cannot be
> 
> simplified to a constant Datum during normalization even if their argument
> 
> is constant.
> 
> 
> 
> Fixes #7039.
> 
> 
> 
> 
> 
> 
</details>


---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[sql/parser/builtins.go, line 665 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7043#-KJP4yO6ta8ByAxhXU64:-KJP4yO7MpsUiVJWIJx8:-1088931579) ([raw file](https://github.com/cockroachdb/cockroach/blob/382976b91000f3a2af55119f6200b790c80f8613/sql/parser/builtins.go#L665)):*
> ```Go
> 	// Aggregate functions.
> 	// These functions are handled in sql/group.go and are not evaluated normally.
> 	//
> ```

We should probably keep some comment about why these are all marked as impure. Something like the comment you deleted for `countImpl`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7043#-:-KJP538eJE2KCMeYyjL-:1417088721)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented on sql/testdata/aggregate:
```diff
@@ -6,6 +6,17 @@ CREATE TABLE kv (
 
+# Aggregate functions return NULL if there are no rows.
+query IIIRRRBB
+SELECT MIN(1), MAX(1), COUNT(1), AVG(1), STDDEV(1), VARIANCE(1), BOOL_AND(true), BOOL_AND(false) FROM kv
```
Aren't we missing SUM here?

knz replied:
Indeed. Moreover, there were two extra use cases missing from the tests too. done.

nvanbenschoten replied:
Yeah, I think the "no source" case was what I had based the previous split off of.

nvanbenschoten commented on sql/testdata/aggregate:
```diff
@@ -6,6 +6,23 @@ CREATE TABLE kv (
+NULL NULL 0 NULL NULL NULL NULL NULL NULL
+
+query IIIRRRRBB
+SELECT MIN(v), MAX(v), COUNT(v), AVG(v), SUM(1), STDDEV(v), VARIANCE(v), BOOL_AND(v = 1), BOOL_AND(v = 1) FROM kv
```
`s/SUM(1)/SUM(v)`?

knz replied:
Absolutely. Thanks for noticing. Fixed.

knz committed:
sql: make all aggregate functions behave consistently.

All aggregate functions except for COUNT should return NULL if the
source table has zero rows, regardless of their argument. This means
they cannot be simplified to a constant Datum during normalization
even if their argument is constant.
